File: 20-actors/fuchi/methods/test_provision.py
#!/usr/bin/env python3
"""Tests for 扶持 (fuchi) provision.py — R1(a) in-kind provisioning intents.

Standalone-runnable: python3 test_provision.py
"""
from __future__ import annotations

import sys

from provision import PROVIDER_REGISTRY, ProvisioningIntent, provision
from route import route_envelope


def _env(line, imputed):
    return {":envelope/line": ":" + line, ":envelope/imputed-usd-micros-yr": imputed,
            ":envelope/cash-usd-micros": 0}


def test_every_rail_has_a_provider():
    rails = route_envelope([_env(l, 1_000_000) for l in
                            ("housing", "food", "energy", "compute", "tooling", "care", "liquidity")])
    intents = provision(rails, "alloc-1")
    assert len(intents) == 7
    providers = {i.provider_did for i in intents}
    assert "did:web:etzhayyim.com:actor:mitsuho" in providers
    assert "commons-land" in providers and "murakumo" in providers


def test_provider_kinds_are_classified():
    rails = route_envelope([_env("housing", 1), _env("food", 1), _env("compute", 1)])
    by = {i.rail_kind: i for i in provision(rails, "a")}
    assert by["housing-commons"].provider_kind == "commons"
    assert by["food-mitsuho"].provider_kind == "actor"
    assert by["compute-murakumo"].provider_kind == "infra"


def test_liquidity_is_member_principal():
    rails = route_envelope([_env("liquidity", 5), _env("food", 5)])
    by = {i.rail_kind: i for i in provision(rails, "a")}
    assert by["liquidity-warifu"].member_principal is True
    assert by["liquidity-warifu"].provider_did.endswith("warifu")
    assert by["food-mitsuho"].member_principal is False


def test_intent_is_dry_run_cashless_keyless():
    for i in provision(route_envelope([_env("food", 1)]), "a"):
        assert i.cash_usd_micros == 0
        assert i.server_held_key is False
        assert i.published is False


def test_published_true_is_refused():
    try:
        ProvisioningIntent("a", "food-mitsuho", "did:..:mitsuho", "actor", 1, published=True)
    except ValueError as e:
        assert "G10" in str(e)
        return
    raise AssertionError("published=true must be refused (G10)")


def test_cash_intent_is_refused():
    try:
        ProvisioningIntent("a", "food-mitsuho", "did:..:mitsuho", "actor", 1, cash_usd_micros=5)
    except ValueError as e:
        assert "cash" in str(e).lower()
        return
    raise AssertionError("nonzero cash intent must be refused (G2)")


def test_server_key_intent_is_refused():
    try:
        ProvisioningIntent("a", "food-mitsuho", "did:..:mitsuho", "actor", 1, server_held_key=True)
    except ValueError as e:
        assert "no-server-key" in str(e)
        return
    raise AssertionError("server-held-key intent must be refused (G9)")


def test_registry_covers_every_route_rail():
    # PROVIDER_REGISTRY must cover exactly the 7 rail kinds
    from route import LINE_TO_RAIL
    rail_kinds = {k for k, _ in LINE_TO_RAIL.values()}
    assert set(PROVIDER_REGISTRY) == rail_kinds


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"test_provision.py: {len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/fuchi/methods/test_provision.clj
;; test_provision.clj — 扶持 provision: rails → provisioning intents (real producing actors) +
;; abaki anti-monopoly route-around, parity with provision.py. Run via `bb test:fuchi`. ADR-2606142300.
(ns fuchi.methods.test-provision
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [fuchi.methods.provision :as p]
            [fuchi.methods.route :as r]
            [fuchi.methods.live-gate :as lg]))

(def ^:private rails
  (r/route-envelope [{:envelope/line :housing   :envelope/imputed-usd-micros-yr 12000000000}
                     {:envelope/line :food      :envelope/imputed-usd-micros-yr 6000000000}
                     {:envelope/line :liquidity :envelope/imputed-usd-micros-yr 2000000000}]))

(deftest provisioning-intents
  (testing "rails → provisioning intents addressed to real producing actors (golden from provision.py)"
    (let [ints (p/provision rails "alloc-1" #{})]            ; no blocked entities
      (is (= 3 (count ints)))
      (is (= [["housing-commons" "commons-land" "commons"]
              ["food-mitsuho" "did:web:etzhayyim.com:actor:mitsuho" "actor"]
              ["liquidity-warifu" "did:web:etzhayyim.com:actor:warifu" "actor"]]
             (mapv (juxt :rail-kind :provider-did :provider-kind) ints)))
      (is (= [false false true] (mapv :member-principal ints)))   ; liquidity = member-principal
      (is (every? #(false? (:published %)) ints))            ; G10 structural
      (is (every? #(= 0 (:cash-usd-micros %)) ints))         ; G2
      (is (every? #(false? (:server-held-key %)) ints)))))   ; G9

(deftest intent-invariants
  (testing "G2 cash≡0 / G9 no-server-key / G10 unpublished / G3 known rail enforced"
    (is (thrown? Exception (p/make-provisioning-intent {:rail-kind "food-mitsuho" :cash-usd-micros 5})))
    (is (thrown? Exception (p/make-provisioning-intent {:rail-kind "food-mitsuho" :server-held-key true})))
    (is (thrown? Exception (p/make-provisioning-intent {:rail-kind "food-mitsuho" :published true})))
    (is (thrown? Exception (p/make-provisioning-intent {:rail-kind "unknown-rail"})))
    (is (map? (p/make-provisioning-intent {:rail-kind "care-iyashi" :imputed-usd-micros-yr 1})))))

(deftest abaki-route-around
  (testing "a provider blocked by the abaki Anti-Monopoly policy raises (route-around)"
    (is (thrown? Exception (p/provision rails "alloc-1" #{"mitsuho"})))   ; mitsuho provider blocked
    (is (= 3 (count (p/provision rails "alloc-1" #{}))))                  ; none blocked → all provisioned
    (is (set? (p/load-blocked-ids "20-actors/abaki/out/does-not-exist.json")))))  ; missing → #{}

(deftest dispatch-gated
  (testing "dispatch-live authorizes via the R2 autonomous gate; cash≡0 / no-server-key hold"
    (let [ints (p/provision rails "alloc-1" #{})
          recs (p/dispatch-live ints (lg/make-gate {:leg "provision"}))]
      (is (= 3 (count recs)))
      (is (every? :authorized-to-publish recs))
      (is (every? #(= 7 (:council-level %)) recs)))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'fuchi.methods.test-provision)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
